apps/mockingbird/mbird.py

```diff
@@ -0,0 +1,905 @@
+        logging.info(f"Updating the output datapoint: {self.name} | {datapoint} | {value}")
+        with varlock:
+            parts = datapoint.split('/')
+            strValue = value.replace("'", '"')
+            zt = f"glp/0/{SS_SID}/rq/dev/lon/{self.name}/if/{parts[0]}/{parts[1]}"
+            zp = "{\"" + parts[2] + "\":{\"value\":" + strValue + ",\"prio\":0}}"
+            logging.debug(f"TOPIC: {zt} | PAYLOAD: {zp}")
+            client.publish(zt, zp)
+
+    def set_health(self, state):
+        logging.info(f"Setting the initial state of {self} to {state}")
+        zt = f"lep/0/lon/0/request/{self.LEP_id}/meta/enabled"
+        if state == "toggle":
+            if self.dev_state == "disabled":
+                state = "true"
+            elif self.dev_state == "provisioned" or self.dev_state == "suspect":
+                state = "false"
+            else:
+                # Default
+                state = "true"
+        # Set the target state.
+        self.target_health = state
+        zp = f"{{\"method\":\"PUT\",\"body\":{state}}}"
+        logging.debug(f"TOPIC: {zt} | PAYLOAD: {zp}")
+        client.publish(zt, zp)
+        
+        return self
+    
+    def service_pin(self):
+        zt = f"lep/0/lon/0/request/cmd/{self.LEP_id}/service/servicePin"
+        zp = "{\"method\":\"PUT\",\"body\":{\"processingState\":\"request\"}}"
+        logging.debug(f"Requested servicePin for {self.name} with LEP ID: {self.LEP_id}")
+        client.publish(zt, zp)
+        return self
+        
+    def __str__(self) -> str:
+        return self.name
+
+
+def on_connect(client, userdata, flags, rc):
+    global connected
+    logging.info("Connected with result code "+str(rc))
+    connected = True
+
+
+def on_disconnect(client, userdata, rc):
+    global connected
+    logging.info("Disconnected with result code "+str(rc))
+    connected = False
+
+
+# create MQTT client object
+client = paho.Client("mbird")     # !! Must be unique !!
+client.on_connect = on_connect
+client.on_disconnect = on_disconnect
+
+# connect
+logging.info('Waiting to connect to {}'.format(args.ip))
+client.connect(args.ip)
+
+# Subscribe to all topics
+client.subscribe(subscribe_topic)
+
+# Register callbacks to specific topics:
+# Callback for obtaining SS SID
+client.message_callback_add(sid_topic, handle_sid)
+# Callback for registering devices:
+# NOTE: make sure that all of the topics using handle_device_creation() callback
+# have a device handle on the message.topic.split("/")[6]
+client.message_callback_add(dev_impl_topic, handle_device_creation)
+client.message_callback_add(dev_sts_topic, handle_device_creation)
+client.message_callback_add(dev_cfg_topic, handle_device_creation)
+# Callback for updating datapoint values
+client.message_callback_add(update_datapoint_topic, handle_datapoint_update)
+
+client.loop_start()
+
+
+if __name__ == "__main__":
+    # Wait for all retaining messages to come from the SmartServer
+    time.sleep(1)
```

Comment:
Is one second long enough when there are 1,000 or even 2,000 devices?

Reply:
BTW, we have had multiple bugs in the past where services were not waiting long enough to get everything.  One thing we did was to require that anything publishing N things would provide the number N somehow in a topic so that listeners knew when they had received it all.  In this case you might not have this.  :(

Reply:
It seems to work just fine like this. Tested with 1000 devices.